consensus/currency.go

```diff
@@ -0,0 +1,114 @@
+package consensus
+
+import (
+	"errors"
+	"math/big"
+)
+
+var (
+	ErrOverflow = errors.New("Currency overflowed 128 bits")
+)
+
+// A Currency is a 128-bit unsigned integer. Currency operations are performed
+// via math/big.
+//
+// The Currency object also keeps track of whether an overflow has occurred
+// during arithmetic operations. Once the 'overflow' flag has been set to
+// true, all subsequent operations will return an error, and the result of the
+// operation is undefined. This flag can never be reset; a new Currency must
+// be created. Callers can also manually check for overflow using the Overflow
+// method.
+type Currency struct {
+	i  big.Int
+	of bool // has an overflow ever occurred?
+}
+
+func NewCurrency(x uint64) Currency {
+	// no possibility of error
+	c, _ := BigToCurrency(new(big.Int).SetUint64(x))
+	return c
+}
+
+func BigToCurrency(b *big.Int) (c Currency, err error) {
+	if b.BitLen() > 128 || b.Sign() < 0 {
+		c.of = true
+		err = ErrOverflow
+		return
+	}
+	c.i = *b
+	return
+}
+
+func (c *Currency) SetBig(b *big.Int) (err error) {
+	oldOF := c.of
+	*c, err = BigToCurrency(b)
+	c.of = c.of || oldOF // preserve overflow flag
+	return
+}
+
+func (c *Currency) Big() *big.Int {
+	return &c.i
+}
+
+func (c *Currency) Add(y Currency) error {
+	if c.of {
+		return ErrOverflow
+	}
+	return c.SetBig(c.i.Add(&c.i, &y.i))
+}
+
+func (c *Currency) Sub(y Currency) error {
+	if c.of {
+		return ErrOverflow
+	}
+	return c.SetBig(c.i.Sub(&c.i, &y.i))
+}
+
+func (c *Currency) Mul(y Currency) error {
+	if c.of {
+		return ErrOverflow
+	}
+	return c.SetBig(c.i.Mul(&c.i, &y.i))
+}
+
+func (c *Currency) Div(y Currency) error {
+	if c.of {
+		return ErrOverflow
+	}
+	return c.SetBig(c.i.Div(&c.i, &y.i))
+}
+
+func (c *Currency) Sqrt() Currency {
+	f, _ := new(big.Rat).SetInt(&c.i).Float64()
+	rat := new(big.Rat).SetFloat64(f)
+	s, _ := BigToCurrency(new(big.Int).Div(rat.Num(), rat.Denom()))
+	s.of = c.of // preserve overflow
+	return s
+}
+
+func (c *Currency) Sign() int {
+	return c.i.Sign()
+}
+
+func (c *Currency) Cmp(y Currency) int {
+	return c.i.Cmp(&y.i)
+}
+
+func (c *Currency) Overflow() bool {
+	return c.of
+}
+
+func (c Currency) MarshalSia() []byte {
+	b := make([]byte, 16)
+	copy(b, c.i.Bytes())
+	return b
+}
+
+func (c *Currency) UnmarshalSia(b []byte) int {
+	var err error
+	*c, err = BigToCurrency(new(big.Int).SetBytes(b[:16]))
+	if err != nil {
+		return -1
+	}
+	return 16
+}
```

Comment:
A negative `big.Int` is encoded just like any other `big.Int`. It'll decode to a negative `big.Int`, which will cause an error.
I guess `Currency` is still technically signed, which means it's only a 127-bit unsigned integer. I think I could fix this by extending it to 129 bits and dropping the leading bit though.

ACTUALLY:
When I investigated further, I found that there's actually no way to create a negative `big.Int` just using `SetBytes`. You to set the negative flag separately. See for yourself: http://play.golang.org/p/kvi76glQ1g

This is great, actually, because it means it's actually impossible to create a negative `Currency`. It's also impossible to create a "negative" `Target`, because `Target.Int` also uses `SetBytes`. So we should be safe.